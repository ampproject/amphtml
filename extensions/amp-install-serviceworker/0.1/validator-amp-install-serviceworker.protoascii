#
# Copyright 2016 The AMP HTML Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS-IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the license.
#

tags: {  # amp-install-serviceworker
  html_format: AMP  # Ads cannot install serviceworkers
  tag_name: "SCRIPT"
  spec_name: "amp-install-serviceworker extension .js script"
  mandatory_parent: "HEAD"
  unique_warning: true
  extension_unused_unless_tag_present: "amp-install-serviceworker"
  attrs: {
    name: "async"
    mandatory: true
    value: ""
  }
  attrs: {
    name: "custom-element"
    value: "amp-install-serviceworker"
    mandatory: true
    dispatch_key: true
  }
  attrs: { name: "nonce" }
  attrs: {
    name: "src"
    mandatory: true
    value_regex: "https://cdn\\.ampproject\\.org/v0/amp-install-serviceworker-(latest|0\\.1).js"
  }
  attrs: {
    name: "type"
    value_casei: "text/javascript"
  }
  cdata: {
    blacklisted_cdata_regex: {
      regex: "."
      error_message: "contents"
    }
  }
  spec_url: "https://www.ampproject.org/docs/reference/extended/amp-install-serviceworker.html"
}
tags: {  # <amp-install-serviceworker>
  html_format: AMP  # Ads cannot install serviceworkers
  tag_name: "AMP-INSTALL-SERVICEWORKER"
  disallowed_ancestor: "AMP-SIDEBAR"
  also_requires_tag: "amp-install-serviceworker extension .js script"
  attrs: {
    name: "src"
    mandatory: true
    value_url: {
      allowed_protocol: "https"
      allow_relative: true  # Will be set to false at a future date.
    }
    blacklisted_value_regex: "__amp_source_origin"
  }
  spec_url: "https://www.ampproject.org/docs/reference/extended/amp-install-serviceworker.html"
  amp_layout: {
    supported_layouts: NODISPLAY
  }
}
