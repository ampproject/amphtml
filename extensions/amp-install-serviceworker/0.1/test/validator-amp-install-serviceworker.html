<!--
  Test Description:
  Tests for the amp-install-serviceworker tag. See the inline comments.
-->
<!doctype html>
<html âš¡ lang="en">
<head>
  <meta charset="utf-8">
  <link rel="canonical" href="./regular-html-version.html">
  <meta name="viewport" content="width=device-width,minimum-scale=1">
  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
  <script async custom-element="amp-install-serviceworker" src="https://cdn.ampproject.org/v0/amp-install-serviceworker-latest.js"></script>
  <script async src="https://cdn.ampproject.org/v0.js"></script>
</head>
<body>
  <!-- Valid: Installs a service worker. -->
  <amp-install-serviceworker
    src="https://www.your-domain.com/serviceworker.js"
    data-iframe-src="https://www.your-domain.com/install-serviceworker.html"
    layout="nodisplay">
  </amp-install-serviceworker>

  <!-- Valid: Has optional data-iframe-src attribute. -->
  <amp-install-serviceworker
    src="https://www.your-domain.com/serviceworker.js"
    layout="nodisplay">
  </amp-install-serviceworker>

  <!-- Invalid: The data-iframe-src attribute is not https or relative. -->
  <amp-install-serviceworker
    src="https://www.your-domain.com/serviceworker.js"
    data-iframe-src="http://www.your-domain.com/install-serviceworker.html"
    layout="nodisplay">
  </amp-install-serviceworker>

  <!-- Invalid: Missing src attribute. -->
  <amp-install-serviceworker
    data-iframe-src="https://www.your-domain.com/install-serviceworker.html"
    layout="nodisplay">
  </amp-install-serviceworker>

  <!-- Invalid: The src attribute is not https or relative. -->
  <amp-install-serviceworker
    src="http://www.your-domain.com/serviceworker.js"
    data-iframe-src="https://www.your-domain.com/install-serviceworker.html"
    layout="nodisplay">
  </amp-install-serviceworker>

  <!-- Invalid: illegal layout. -->
  <amp-install-serviceworker
    src="https://www.your-domain.com/serviceworker.js"
    data-iframe-src="https://www.your-domain.com/install-serviceworker.html"
    layout="responsive">
  </amp-install-serviceworker>
</body>
</html>
