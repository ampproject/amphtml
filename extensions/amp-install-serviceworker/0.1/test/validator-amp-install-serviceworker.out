FAIL
|  <!--
|    Copyright 2018 The AMP HTML Authors. All Rights Reserved.
|
|    Licensed under the Apache License, Version 2.0 (the "License");
|    you may not use this file except in compliance with the License.
|    You may obtain a copy of the License at
|
|        http://www.apache.org/licenses/LICENSE-2.0
|
|    Unless required by applicable law or agreed to in writing, software
|    distributed under the License is distributed on an "AS-IS" BASIS,
|    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
|    See the License for the specific language governing permissions and
|    limitations under the license.
|  -->
|  <!--
|    Test Description:
|    Tests for the amp-install-serviceworker tag. See the inline comments.
|  -->
|  <!doctype html>
|  <html âš¡>
|  <head>
|    <meta charset="utf-8">
|    <link rel="canonical" href="./regular-html-version.html">
|    <meta name="viewport" content="width=device-width,minimum-scale=1">
|    <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
|    <script async custom-element="amp-install-serviceworker" src="https://cdn.ampproject.org/v0/amp-install-serviceworker-latest.js"></script>
|    <script async src="https://cdn.ampproject.org/v0.js"></script>
|  </head>
|  <body>
|    <!-- Valid: Installs a service worker. -->
|    <amp-install-serviceworker
|      src="https://www.your-domain.com/serviceworker.js"
|      data-iframe-src="https://www.your-domain.com/install-serviceworker.html"
|      layout="nodisplay">
|    </amp-install-serviceworker>
|
|    <!-- Valid: Has optional data-iframe-src attribute. -->
|    <amp-install-serviceworker
|      src="https://www.your-domain.com/serviceworker.js"
|      layout="nodisplay">
|    </amp-install-serviceworker>
|
|    <!-- Invalid: The data-iframe-src attribute is not https or relative. -->
|    <amp-install-serviceworker
>>   ^~~~~~~~~
amp-install-serviceworker/0.1/test/validator-amp-install-serviceworker.html:45:2 Invalid URL protocol 'http:' for attribute 'data-iframe-src' in tag 'amp-install-serviceworker'. (see https://www.ampproject.org/docs/reference/components/amp-install-serviceworker) [AMP_TAG_PROBLEM]
|      src="https://www.your-domain.com/serviceworker.js"
|      data-iframe-src="http://www.your-domain.com/install-serviceworker.html"
|      layout="nodisplay">
|    </amp-install-serviceworker>
|
|    <!-- Invalid: Missing src attribute. -->
|    <amp-install-serviceworker
>>   ^~~~~~~~~
amp-install-serviceworker/0.1/test/validator-amp-install-serviceworker.html:52:2 The mandatory attribute 'src' is missing in tag 'amp-install-serviceworker'. (see https://www.ampproject.org/docs/reference/components/amp-install-serviceworker) [AMP_TAG_PROBLEM]
|      data-iframe-src="https://www.your-domain.com/install-serviceworker.html"
|      layout="nodisplay">
|    </amp-install-serviceworker>
|
|    <!-- Invalid: The src attribute is not https or relative. -->
|    <amp-install-serviceworker
>>   ^~~~~~~~~
amp-install-serviceworker/0.1/test/validator-amp-install-serviceworker.html:58:2 Invalid URL protocol 'http:' for attribute 'src' in tag 'amp-install-serviceworker'. (see https://www.ampproject.org/docs/reference/components/amp-install-serviceworker) [AMP_TAG_PROBLEM]
|      src="http://www.your-domain.com/serviceworker.js"
|      data-iframe-src="https://www.your-domain.com/install-serviceworker.html"
|      layout="nodisplay">
|    </amp-install-serviceworker>
|
|    <!-- Invalid: illegal layout. -->
|    <amp-install-serviceworker
>>   ^~~~~~~~~
amp-install-serviceworker/0.1/test/validator-amp-install-serviceworker.html:65:2 The specified layout 'RESPONSIVE' is not supported by tag 'amp-install-serviceworker'. (see https://www.ampproject.org/docs/reference/components/amp-install-serviceworker) [AMP_LAYOUT_PROBLEM]
|      src="https://www.your-domain.com/serviceworker.js"
|      data-iframe-src="https://www.your-domain.com/install-serviceworker.html"
|      layout="responsive">
|    </amp-install-serviceworker>
|  </body>
|  </html>
